In [5]:
import pandas as pd
import numpy as np
import seaborn as sns

In [8]:
#membaca file csv
data = pd.read_csv('transaction_data.csv', engine='python')

#menampilkan beberapa baris pertama
print(data.head())

   UserId  TransactionId               TransactionTime  ItemCode  \
0  278166        6355745  Sat Feb 02 12:50:00 IST 2019    465549   
1  337701        6283376  Wed Dec 26 09:06:00 IST 2018    482370   
2  267099        6385599  Fri Feb 15 09:45:00 IST 2019    490728   
3  380478        6044973  Fri Jun 22 07:14:00 IST 2018    459186   
4      -1        6143225  Mon Sep 10 11:58:00 IST 2018   1733592   

                     ItemDescription  NumberOfItemsPurchased  CostPerItem  \
0   FAMILY ALBUM WHITE PICTURE FRAME                       6        11.73   
1              LONDON BUS COFFEE MUG                       3         3.52   
2  SET 12 COLOUR PENCILS DOLLY GIRL                       72         0.90   
3        UNION JACK FLAG LUGGAGE TAG                       3         1.73   
4                WASHROOM METAL SIGN                       3         3.40   

          Country  
0  United Kingdom  
1  United Kingdom  
2          France  
3  United Kingdom  
4  United Kingdom  


perlu mengetahui kondisi data, dengan memeriksa data yang kosong.

In [9]:
# Bersihkan data dari nilai yang hilang atau tidak konsisten
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1083818 entries, 0 to 1083817
Data columns (total 8 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   UserId                  1083818 non-null  int64  
 1   TransactionId           1083818 non-null  int64  
 2   TransactionTime         1083818 non-null  object 
 3   ItemCode                1083818 non-null  int64  
 4   ItemDescription         1080910 non-null  object 
 5   NumberOfItemsPurchased  1083818 non-null  int64  
 6   CostPerItem             1083818 non-null  float64
 7   Country                 1083818 non-null  object 
dtypes: float64(1), int64(4), object(3)
memory usage: 66.2+ MB


In [10]:
# mengecek nilai NaN atau NULL
data.isnull().sum()

,0
UserId,0
TransactionId,0
TransactionTime,0
ItemCode,0
ItemDescription,2908
NumberOfItemsPurchased,0
CostPerItem,0
Country,0


In [17]:
# menghapus baris yang mengandung nilai NaN atau NULL
data.dropna(inplace=True)
data.info()


<class 'pandas.core.frame.DataFrame'>
Index: 1080910 entries, 0 to 1083817
Data columns (total 8 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   UserId                  1080910 non-null  int64  
 1   TransactionId           1080910 non-null  int64  
 2   TransactionTime         1080910 non-null  object 
 3   ItemCode                1080910 non-null  int64  
 4   ItemDescription         1080910 non-null  object 
 5   NumberOfItemsPurchased  1080910 non-null  int64  
 6   CostPerItem             1080910 non-null  float64
 7   Country                 1080910 non-null  object 
dtypes: float64(1), int64(4), object(3)
memory usage: 74.2+ MB


Setelah menghilangkan data hilang, sekarang data menjadi 1080910 entri. Maka bisa dikatakan setidaknya terdapat satu nilai yang hilang. Untuk mengidentifikasi pola pembelian barang oleh pelanggan dan merumuskan stategi pemasaran berdasarkan aturan asosiasi yang dihasilkan. Pertama perlu membuat agregat berdasarkan TransactionId untuk mendapatkan daftar barang per transaks

In [15]:
#agrerat TransactionId

basketplus = data.groupby(['TransactionId','ItemDescription'])['NumberOfItemsPurchased'].sum().unstack().reset_index().fillna(0).set_index('TransactionId')
basketplus

ItemDescription,4 PURPLE FLOCK DINNER CANDLES,50'S CHRISTMAS GIFT BAG LARGE,DOLLY GIRL BEAKER,I LOVE LONDON MINI BACKPACK,I LOVE LONDON MINI RUCKSACK,NINE DRAWER OFFICE TIDY,OVAL WALL MIRROR DIAMANTE,RED SPOT GIFT BAG LARGE,SET 2 TEA TOWELS I LOVE LONDON,SPACEBOY BABY GIFT SET,...,wrongly coded 20713,wrongly coded 23343,wrongly coded-23343,wrongly marked,wrongly marked 23343,wrongly marked carton 22804,wrongly marked. 23343 in box,wrongly sold (22719) barcode,wrongly sold as sets,wrongly sold sets
TransactionId,,,,,,,,,,,,,,,,,,,,,
5900015,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
5900026,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
5900037,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
5900048,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
5900059,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6397413,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
6397424,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
6397435,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Dalam mengidentifikasi pola pembelian barang oleh pelanggan, penting mengetahui suatu barang terbeli atau tidak. Sebab, kita hanya ingin mengetahui, apa hubungan antara membeli beberapa barang dan membeli beberapa barang lainnya. Jadi, kita perlu mengkodekan data keranjang menjadi data biner yang menunjukkan apakah suatu barang dibeli (1) atau tidak (0).

In [16]:
# Kode Beli Barang

def encode_units(x):
  if x<=0:
    return 0
  if x>= 1:
    return 1

basket_encode_plus = basketplus.applymap(encode_units)
basket_encode_plus


/tmp/ipykernel_565/2228233064.py:9: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encode_plus = basketplus.applymap(encode_units)


ItemDescription,4 PURPLE FLOCK DINNER CANDLES,50'S CHRISTMAS GIFT BAG LARGE,DOLLY GIRL BEAKER,I LOVE LONDON MINI BACKPACK,I LOVE LONDON MINI RUCKSACK,NINE DRAWER OFFICE TIDY,OVAL WALL MIRROR DIAMANTE,RED SPOT GIFT BAG LARGE,SET 2 TEA TOWELS I LOVE LONDON,SPACEBOY BABY GIFT SET,...,wrongly coded 20713,wrongly coded 23343,wrongly coded-23343,wrongly marked,wrongly marked 23343,wrongly marked carton 22804,wrongly marked. 23343 in box,wrongly sold (22719) barcode,wrongly sold as sets,wrongly sold sets
TransactionId,,,,,,,,,,,,,,,,,,,,,
5900015,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900026,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900037,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900048,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900059,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6397413,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
6397424,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
6397435,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


Kemudian perlu mengetahui hubungan antara 2 atau lebih item yang dibeli berdarkan data yang ada.  Jadi, kurang bermanfaat jika transaksi hanya membeli satu item saja.  Oleh karena itu, langkah selanjutnya adalah menyaring transaksi pembelian lebih dari 1 item.

In [18]:
#Filter

basket_filter_plus = basket_encode_plus[(basket_encode_plus > 0).sum(axis=1) >= 2]
basket_filter_plus


ItemDescription,4 PURPLE FLOCK DINNER CANDLES,50'S CHRISTMAS GIFT BAG LARGE,DOLLY GIRL BEAKER,I LOVE LONDON MINI BACKPACK,I LOVE LONDON MINI RUCKSACK,NINE DRAWER OFFICE TIDY,OVAL WALL MIRROR DIAMANTE,RED SPOT GIFT BAG LARGE,SET 2 TEA TOWELS I LOVE LONDON,SPACEBOY BABY GIFT SET,...,wrongly coded 20713,wrongly coded 23343,wrongly coded-23343,wrongly marked,wrongly marked 23343,wrongly marked carton 22804,wrongly marked. 23343 in box,wrongly sold (22719) barcode,wrongly sold as sets,wrongly sold sets
TransactionId,,,,,,,,,,,,,,,,,,,,,
5900015,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900026,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900037,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900048,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
5900070,0,0,0,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6397413,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
6397424,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
6397435,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


Berdasarkan hasil di atas terlihat terdapat 18.338 transaksi yang membeli lebih dari satu item. Setelah menghasilkan dataset tersebut, tahap selanjutnya adalah menerapkan algoritma Apriori. Algoritma Apriori digunakan untuk menemukan frequent itemset, yaitu item atau kombinasi item yang sering muncul secara bersamaan dalam transaksi. Pada penelitian ini, nilai minimum support (min_support) ditetapkan sebesar 0,03, sehingga hanya item atau kombinasi item yang muncul pada minimal 3% dari keseluruhan transaksi yang akan dianggap sebagai frequent itemset. Penentuan nilai tersebut bertujuan untuk menyaring item yang cukup sering muncul sehingga pola yang dihasilkan lebih representatif dan tidak menghasilkan terlalu banyak maupun terlalu sedikit kombinasi item.

In [19]:
#Apriori

from mlxtend.frequent_patterns import apriori

frequent_itemsets_plus = apriori(basket_filter_plus, min_support=0.03,
                                 use_colnames=True).sort_values('support', ascending=False).reset_index(drop=True)

frequent_itemsets_plus['length'] = frequent_itemsets_plus['itemsets'].apply(lambda x: len(x))

frequent_itemsets_plus


/usr/local/lib/python3.12/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: Dep

,support,itemsets,length
0,0.122205,(WHITE HANGING HEART T-LIGHT HOLDER),1
1,0.113317,(JUMBO BAG RED RETROSPOT),1
2,0.107264,(REGENCY CAKESTAND 3 TIER),1
3,0.091122,(PARTY BUNTING),1
4,0.085233,(LUNCH BAG RED RETROSPOT),1
...,...,...,...
160,0.030320,"(LUNCH BAG RED RETROSPOT, LUNCH BAG SUKI DESIGN )",2
161,0.030210,"(LUNCH BAG CARS BLUE, LUNCH BAG RED RETROSPOT)",2
162,0.030210,(HAND WARMER BIRD DESIGN),1
163,0.030101,(CHILDRENS APRON SPACEBOY DESIGN),1


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


Algoritma Apriori digunakan untuk menemukan frequent itemset, yaitu item atau kombinasi item yang sering muncul secara bersamaan dalam transaksi berdasarkan nilai minimum support yang telah ditentukan.

Length = 1 → menunjukkan 1 item yang sering dibeli.
Contoh: WHITE HANGING HEART T-LIGHT HOLDER memiliki support 0,122205, artinya produk tersebut muncul pada sekitar 12,22% dari seluruh transaksi.

Length = 2 → menunjukkan kombinasi 2 item yang sering dibeli secara bersamaan.
Contoh: LUNCH BAG SUKI DESIGN dan LUNCH BAG RED RETROSPOT muncul bersama pada sekitar 3,03% transaksi.


Setelah menerapkan algoritma apriori dan menemukan barang yang sering dibeli, kini dapat menerapkan aturan asosiasi. Dari aturan asosiasi, kita bisa mengetahui informasi bahkan menemukan pengetahuan tentang barang mana yang lebih efektif untuk dijual bersama.


In [20]:
# Asosiasi
from mlxtend.frequent_patterns import association_rules

association_rules(frequent_itemsets_plus, metric='lift',
                  min_threshold=1).sort_values('lift', ascending=False).reset_index(drop=True)


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(PINK REGENCY TEACUP AND SAUCER),(GREEN REGENCY TEACUP AND SAUCER),0.041499,0.055350,0.034518,0.831800,15.028131,1.0,0.032222,5.616242,0.973872,0.553806,0.821945,0.727723
1,(GREEN REGENCY TEACUP AND SAUCER),(PINK REGENCY TEACUP AND SAUCER),0.055350,0.041499,0.034518,0.623645,15.028131,1.0,0.032222,2.546804,0.988152,0.553806,0.607351,0.727723
2,(ROSES REGENCY TEACUP AND SAUCER ),(PINK REGENCY TEACUP AND SAUCER),0.057967,0.041499,0.032664,0.563500,13.578784,1.0,0.030259,2.195877,0.983358,0.488980,0.544601,0.675311
3,(PINK REGENCY TEACUP AND SAUCER),(ROSES REGENCY TEACUP AND SAUCER ),0.041499,0.057967,0.032664,0.787122,13.578784,1.0,0.030259,4.425229,0.966462,0.488980,0.774023,0.675311
4,(ROSES REGENCY TEACUP AND SAUCER ),(GREEN REGENCY TEACUP AND SAUCER),0.057967,0.055350,0.041880,0.722484,13.053107,1.0,0.038672,3.403944,0.980210,0.586260,0.706223,0.739567
5,(GREEN REGENCY TEACUP AND SAUCER),(ROSES REGENCY TEACUP AND SAUCER ),0.055350,0.057967,0.041880,0.756650,13.053107,1.0,0.038672,3.871107,0.977494,0.586260,0.741676,0.739567
6,(ALARM CLOCK BAKELIKE RED ),(ALARM CLOCK BAKELIKE GREEN),0.057258,0.053168,0.034900,0.609524,11.464049,1.0,0.031856,2.424813,0.968209,0.462094,0.587597,0.632967
7,(ALARM CLOCK BAKELIKE GREEN),(ALARM CLOCK BAKELIKE RED ),0.053168,0.057258,0.034900,0.656410,11.464049,1.0,0.031856,2.743801,0.964026,0.462094,0.635542,0.632967
8,(PAPER CHAIN KIT 50'S CHRISTMAS ),(PAPER CHAIN KIT VINTAGE CHRISTMAS),0.062875,0.044498,0.030101,0.478751,10.758992,1.0,0.027304,1.833102,0.967912,0.389555,0.454477,0.577611
9,(PAPER CHAIN KIT VINTAGE CHRISTMAS),(PAPER CHAIN KIT 50'S CHRISTMAS ),0.044498,0.062875,0.030101,0.676471,10.758992,1.0,0.027304,2.896568,0.949296,0.389555,0.654764,0.577611


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


Asosiasi adalah aturan yang menunjukkan hubungan antaritem dalam transaksi, yaitu jika pelanggan membeli suatu item (antecedent), maka ada kecenderungan pelanggan juga membeli item lain (consequent).

Dari hasil asosiasi diperoleh bahwa item Green Regency Teacup And Saucer dan Pink Regency Teacup And Saucer sangat bagus untuk dijual secara bersamaan. Karena pada dua item tersebut memiliki asosiasi paling tinggi dengan nilai lift yang paling tinggi, yaitu sebesar 15.028131. Semakin tinggi nilai lift, semakin tinggi pula hubungan antar item. Jika nilai lift lebih dari satu, maka cukup dikatakan bahwa kedua item tersebut saling berhubungan. Kemudian dari hasil asosiasi item Green Regency Teacup And Saucer dan Pink Regency Teacup And Saucer diperoleh nilai support sebesar 0.34518. Dapat diartikan terdapat 3.4518% dari total transaksi yang menjual kedua item tersebut secara bersamaan.
Perlu diketahui bahwa jika antesedennya lebih tinggi dari konsekuennya, maka aturan yang diterapkan adalah aturan nomor 1 dan sebaliknya. Dalam hal ini, nilai antesedennya lebih tinggi dibandingkan nilai konsekuensinya. Artinya dapat menerapkan aturan nomor 1 yaitu Green Regency Teacup And Saucer → Pink Regency Teacup And Saucer. Dapat dijelaskan bahwa pelanggan akan cenderung membeli Pink Regency Teacup And Saucer setelah membeli Green Regency Teacup And Saucer.

Hasil analisis keranjang pasar ini dapat digunakan untuk strategi pemasaran dan pengambilan keputusan berdasarkan data. Dalam kasus ini, dapat memperoleh strategi sebagai berikut.

1.	Membuat item Green Regency Teacup And Saucer dan Pink Regency Teacup And Saucer sebagai satu paket produk dengan harga yang lebih rendah dibandingkan dengan gabungan masing-masing harga. Cara ini akan menarik lebih banyak penjualan dan menghasilkan lebih banyak pendapatan.
2.	Menempatkan item Green Regency Teacup And Saucer dan Pink Regency Teacup And Saucer di tempat yang lebih dekat, atau mungkin di rak yang sama.
3.	Memberikan diskon untuk Green Regency Teacup And Saucer jika pelanggan membeli Pink Regency Teacup And Saucer.
4.	Dapat meletakkan di kasir, sehingga setiap kali ada pelanggan yang membeli Green Regency Teacup And Saucer, dapat menawarkan dan merekomendasikan mereka untuk membeli Cangkir Teh dan Pink Regency Teacup And Saucer dengan harga yang lebih murah.